src/main/java/uk/gov/companieshouse/api/accounts/service/impl/SecretaryServiceImpl.java

```diff
@@ -0,0 +1,181 @@
+package uk.gov.companieshouse.api.accounts.service.impl;
+
+import com.mongodb.MongoException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.GenerateEtagUtil;
+import uk.gov.companieshouse.api.accounts.Kind;
+import uk.gov.companieshouse.api.accounts.ResourceName;
+import uk.gov.companieshouse.api.accounts.exception.DataException;
+import uk.gov.companieshouse.api.accounts.links.BasicLinkType;
+import uk.gov.companieshouse.api.accounts.links.DirectorsReportLinkType;
+import uk.gov.companieshouse.api.accounts.model.entity.directorsreport.SecretaryEntity;
+import uk.gov.companieshouse.api.accounts.model.rest.directorsreport.Secretary;
+import uk.gov.companieshouse.api.accounts.repository.SecretaryRepository;
+import uk.gov.companieshouse.api.accounts.service.ResourceService;
+import uk.gov.companieshouse.api.accounts.service.response.ResponseObject;
+import uk.gov.companieshouse.api.accounts.service.response.ResponseStatus;
+import uk.gov.companieshouse.api.accounts.transformer.SecretaryTransformer;
+import uk.gov.companieshouse.api.accounts.utility.impl.KeyIdGenerator;
+import uk.gov.companieshouse.api.model.transaction.Transaction;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Service
+public class SecretaryServiceImpl implements ResourceService<Secretary> {
+
+    private SecretaryTransformer transformer;
+    private SecretaryRepository secretaryRepository;
+    private DirectorsReportServiceImpl directorsReportService;
+    private KeyIdGenerator keyIdGenerator;
+
+    @Autowired
+    public SecretaryServiceImpl(SecretaryTransformer transformer, SecretaryRepository secretaryRepository, DirectorsReportServiceImpl directorsReportService, KeyIdGenerator keyIdGenerator) {
+
+        this.transformer = transformer;
+        this.secretaryRepository = secretaryRepository;
+        this.directorsReportService = directorsReportService;
+        this.keyIdGenerator = keyIdGenerator;
+    }
+
+    private static final Pattern SECRETARY_ID_REGEX = Pattern.compile("^/transactions/.+?/company-accounts/.+?/small-full/directors-report/secretaries/(.*)$");
+
+
+    @Override
+    public ResponseObject<Secretary> create(Secretary rest, Transaction transaction, String companyAccountId, HttpServletRequest request) throws DataException {
+
+        String secretaryId = keyIdGenerator.generateRandom();
```

Comment:
For this resource, since we now don't have multiple 'secretaries' we can now generate a consistent id by hashing companyAccountId with a resource name as we do in other services (see tangible for example). This will be useful for a number of reasons, but it also removes the need for regexing the request URI in this class